README
=== introduction

An experimental Gearmand distributed server written in Erlang with all the things Gearman should not have: replicated persistent queues,
incipient error recovery and support for clustering.

=== State of the implementation

You can take a look at the CHANGELOG file to see the state of the implementation of the Gearman protocol.
At this moment, the binary and administrative protocol are fully implemented.
The server has been tested with the java and ruby Gearman client libraries using a single node configuration.

=== dependencies

Erlang (tested with OTP R13B01),
Ruby and Rake for compiling

=== OTP application support

The application always start as an OTP application using the egearmand script , you can
start the shell with:
> application:start(egearmand) .
You can change the configuration of the application editing the
file ebin/egearmand.app, editing the host, port, and log mechanism.

=== extensions

From version 0.3 the server has initial support for extensions.
An extension connecting the Gearman server implementation to the
RabbitMQ queue system is included. To use the extension, be sure
you have erlang-rfc4627 in the path and before calling to
gearmand:start(), invoke rabbitmq_extension:start() from the
erlang shell.

The dependencies for building the extension are:

-rabbitmq server
-rabbitmq-erlang-client
-erlang-rfc4627 (included in the contrib directory)

You can find instructions about setting up rabbitmq-server and the erlang client
on OS X here: http://antoniogarrote.lacoctelera.net/post/2009/08/27/installing-rabbitmq-erlang-client-in-os-x

There are examples in the /examples directory  of three ruby
scripts creating, publishing and consuming messages from a queue
through the egearmand server.

=== tests

Few and scattered through the source files.
To run all the tests, starts the erlang shell with the loaded code of the server and run:
> tests:test() .

=== configuration

The default configuration: host, port, log level and log file can be configured in the egearmand.app
application specification file or can be provided as options to the egearmand boot script.

Other configuration options regarding persisten queues and extension can only we configured editing the
file src/configuration.erl. Take a look at the comments.

It must be edited before compiling.

If you want to use persistent queues, you must create the replicated Mnesia schema before calling to gearmand:start().
You can use mnesia_store:create_tables(), after initiating calling to mnesia:create_schema(configuration:gearmand_nodes()).

=== compile

$rake

Optionally, if you want to build the extensions (take into account that you must meet the required dependencies) edit
src/configuration.erl adding the desired extensions and try:

$rake extensions

=== running

./egearmand [-host host] [-port port] [-log path] [-level (debug | info | warning | error)]

=== examples

The distribution includes some examples of clients and workers using to test the server and the RabbitMQ extension.
The examples are coded in Ruby, and use the xing-gearman-ruby gem to work. You can obtain it from github: git://github.com/xing/gearman-ruby.git

=== clustering

A cluster of Egearmand servers is formed by a master node and a set of slave nodes. The file egearmand.app must be edited in the master node
to state the node names of the slave nodes, the slave parameter in that file must be set to false. For instance:

{application,egearmand,
 [{description,"gearman server"},
  {vsn,"0.6.5"},
  {modules,[client_proxy, configuration, connections, functions_registry, gearmand, gearmand_supervisor,
            jobs_queue_server, lists_extensions, log, mnesia_store, poplists_extensions, protocol, mnesia_store, tests, worker_proxy, administration]},
  {registered,[inets_sup, httpc_manager]},
  {applications,[kernel,stdlib, mnesia]},
  {env, [{nodes, ['slave@laptop.local']}]},
  {mod,{egearmand_app,[ {host, "localhost"},
                        {port, 4730},
                        {level, info},
                        {method, file},
                        {path, "egearmand.log"},
                        {check_nodes, false},
                        {slave, false}]}}]}.


This .app file states that a master serve will be started in this node and that a slave node must be started in the node 'vbox@laptop.local'.

In the slave node, the .app file must be edited with an empty list of nodes and the slave parameter to true:

{application,egearmand,
 [{description,"gearman server"},
  {vsn,"0.6.5"},
  {modules,[client_proxy, configuration, connections, functions_registry, gearmand, gearmand_supervisor,
            jobs_queue_server, lists_extensions, log, mnesia_store, poplists_extensions, protocol, mnesia_store, tests, worker_proxy, administration]},
  {registered,[inets_sup, httpc_manager]},
  {applications,[kernel,stdlib, mnesia]},
  {env, [{nodes, []}]},
  {mod,{egearmand_app,[ {host, "localhost"},
                        {port, 4730},
                        {level, info},
                        {method, file},
                        {path, "egearmand.log"},
                        {check_nodes, true},
                        {slave, true}]}}]}.

Now, in each machine where the slaves nodes are going to be executed, an Erlang node with the correct full name must be started:

laptop$ erl -pa path/to/egearmand/ebin -name vbox@laptop.local -setcookie test

Once all the slave nodes are running, the master node can be started with the egearmand script as usual:

server$ ./egearmand -level debug -name master@server.local -setcookie test

The master node will start the supervisor tree for the slave instances of egearmand.

CHANGELOG
=== version 0.6.6 (02/12/2009)

- missing protocol messages
- administrative connections remain open
- proper cleanup of workers after disconnection
- implemented missing status_req/status_res commands
- tested using gearman C version 0.11 examples
- multiple jobs assigned to a single client proxy fixes

=== version 0.6.5 (18/10/2009)

- initial support for automatically start a cluster of gearman servers.

=== version 0.6.4 (11/10/2009)

- splitted rabbitmq extension from main code of the server. Now it is possible to build
  the server withouth the rabbitmq extensions and its dependencies

=== version 0.6.3 (10/10/2009)

- full administrative protocol implemented: workers, status, shutdown, version, maxqueue

=== version 0.6.2 (09/10/2009)

- added global register of worker proxies in order to finish implementation of
  the administrative protocol and clean shutdown of the server.

=== version 0.6.1 (08/10/2009)

- unified boot of the server as an OTP application, also from the egearmand script

=== version 0.6.0 (06/10/2009)

- initial support for the administrative protocol: status, version and shutdown commands

=== version 0.5.1 (04/10/2009)

- fixed version in egearmand.app
- fixed error with set_id requests for the workers
- fixed error with work_fail rqeuests from the workers

=== version 0.5 (09/06/2009)

- OTP compliant application: supervisors, .app file, etc.
- logger server
- updated scripts for starting the server from command line
- improved compilation scripts

=== version 0.4 (09/06/2009)

- Changelog
- Support for all messages in the Gearman binary protocol.
- Improved documentation and logging.
- Centralized tests execution


src/protocol.erl
-module(protocol) .

-author("Antonio Garrote Hernandez") .

-include_lib("eunit/include/eunit.hrl").

%-export([parse_command/2, process_request/2, parse_command/1]) .
-compile(export_all).

%% @doc
%% @author 'Samuel Stauffer <samuel@lefora.com>'
%% @see git://github.com/samuel/erlang-gearman.git
-spec(parse_command/1 :: (Data :: binary()) -> {'error', 'not_enough_data'} | {'ok', binary(), atom(), atom()} | {'ok', binary(), atom(), tuple()}).
parse_command(<<"status", _Rest/binary>>) ->
    {ok, <<>>, administrative, status} ;

parse_command(<<"workers", _Rest/binary>>) ->
    {ok, <<>>, administrative, workers} ;

parse_command(<<"shutdown", _Rest/binary>>) ->
    {ok, <<>>, administrative, shutdown} ;

parse_command(<<"version", _Rest/binary>>) ->
    {ok, <<>>, administrative, version} ;

parse_command(<<"maxqueue", _Rest/binary>>) ->
    {ok, <<>>, administrative, maxqueue} ;

parse_command(Data) when size(Data) >= 12 ->
    <<"\000RE", TypeChar:8, CommandID:32/big, DataLength:32/big, Rest/binary>> = Data,
    if
        size(Rest) >= DataLength ->
            Type = case TypeChar of
                $Q -> request;
                $S -> response
            end,
            {ArgData, LeftOverData} = split_binary(Rest, DataLength),
            Command = parse_command(CommandID, binary_to_list(ArgData)),
            {ok, LeftOverData, Type, Command};
        true ->
            {error, not_enough_data}
    end;
parse_command(_Data) ->
    {error, not_enough_data}.

parse_command(CommandID, Data) ->
    case CommandID of
        22 -> {set_client_id, Data}; % ClientID
        16 -> {echo_req, Data}; % Text
        12 ->
            [Handle, Numerator, Denominator] = split(Data, 0, 2),
            {work_status, Handle, list_to_integer(Numerator), list_to_integer(Denominator)};
        13 ->
            [Handle, Result] = split(Data, 0, 1),
            {work_complete, Handle, Result};
        28 ->
            [Handle, Opaque] = split(Data, 0, 1),
            {work_data, Handle, Opaque};
        29 ->
            [Handle, Opaque] = split(Data, 0, 1),
            {work_warning, Handle, Opaque};
        25 ->
            [Handle, Opaque] = split(Data, 0, 1),
            {work_exception, Handle, Opaque};
        14 -> {work_fail, Data}; % Handle
        17 -> {echo_res, Data}; % Text
        19 ->
            [Code, Text] = split(Data, 0, 1),
            {error, list_to_integer(Code), Text};
        1 -> {can_do, Data}; % Function
        23 ->
            [Function, Timeout] = split(Data, 0, 1),
            {can_do_timeout, Function, list_to_integer(Timeout)};
        2 -> {cant_do, Data}; % Function
        3 -> reset_abilities;
        4 -> pre_sleep;
        9 -> grab_job;
        30 -> grab_job_uniq;
        11 ->
            [Handle, Function, Argument] = split(Data, 0, 2),
            {job_assign, Handle, Function, Argument};
        31 ->
            [Handle, Function, UniqueId, Argument] = split(Data, 0, 2),
            {job_assign_uniq, Handle, Function, UniqueId, Argument};
        24 -> all_yours;
        6 -> noop;
        10 -> no_job;
        7 ->
            [Function, Unique, Argument] = split(Data, 0, 2),
            {submit_job, Function, Unique, Argument};
        21 ->
            [Function, Unique, Argument] = split(Data, 0, 2),
            {submit_job_high, Function, Unique, Argument};
        32 ->
            [Function, Unique, Argument] = split(Data, 0, 2),
            {submit_job_high_bg, Function, Unique, Argument};

        33 ->
            [Function, Unique, Argument] = split(Data, 0, 2),
            {submit_job_low, Function, Unique, Argument};

        34 ->
            [Function, Unique, Argument] = split(Data, 0, 2),
            {submit_job_low_bg, Function, Unique, Argument};

        18 ->
            [Function, Unique, Argument] = split(Data, 0, 2),
            {submit_job_bg, Function, Unique, Argument};
        8 -> {job_created, Data}; % Handle
        15 -> {get_status, Data}; % Handle
        26 -> {option_req, Data}; % Name of the option
        27 -> {option_res, Data}; % Name of the option
        20 ->
            [Handle, Known, Running, Numerator, Denominator] = split(Data, 0, 4),
            {status_res, Handle, list_to_integer(Known), list_to_integer(Running), list_to_integer(Numerator), list_to_integer(Denominator)} ;
        Other -> log:error(["protocol : unknown message ", Other]),
                 noexisto:functionon(Other)
    end.

pack_request(Command, Args) when is_atom(Command), is_tuple(Args) ->
    {CommandID, ArgList} = pack_command(Command, Args),
    pack_command(CommandID, ArgList, "\000REQ").
pack_response(Command, Args) when is_atom(Command), is_tuple(Args) ->
    {CommandID, ArgList} = pack_command(Command, Args),
    pack_command(CommandID, ArgList, "\000RES").
pack_command(CommandID, Args, Magic) when is_integer(CommandID), is_list(Args), is_list(Magic) ->
    Data = list_to_binary(join(Args, [0])),
    DataLength = size(Data),
    list_to_binary([Magic, <<CommandID:32/big, DataLength:32/big>>, Data]).

pack_command(set_client_id, {ClientID}) -> {22, [ClientID]};
pack_command(echo_req, {Text}) -> {16, [Text]};
pack_command(work_status, {Handle, Numerator, Denominator}) -> {12, [Handle, integer_to_list(Numerator), integer_to_list(Denominator)]};
pack_command(work_complete, {Handle, Result}) -> {13, [Handle, Result]};
pack_command(work_data, {Handle, Result}) -> {28, [Handle, Result]};
pack_command(work_warning, {Handle, Warning}) -> {29, [Handle, Warning]};
pack_command(work_fail, {Handle}) -> {14, [Handle]};
pack_command(work_exception, {Handle, Reason}) -> {25, [Handle, Reason]};
pack_command(echo_res, {Text}) -> {17, [Text]};
pack_command(error, {Code, Text}) -> {19, [integer_to_list(Code), Text]};
pack_command(can_do, {Func}) -> {1, [Func]};
pack_command(can_do_timeout, {Func, Timeout}) -> {23, [Func, integer_to_list(Timeout)]};
pack_command(cant_do, {Func}) -> {2, [Func]};
pack_command(reset_abilities, {}) -> {3, []};
pack_command(pre_sleep, {}) -> {4, []};
pack_command(grab_job, {}) -> {9, []};
pack_command(grab_job_uniq, {}) -> {30, []};
pack_command(job_assign, {Handle, Func, Arg}) -> {11, [Handle, Func, Arg]};
pack_command(job_assign_uniq, {Handle, Func, UniqueId, Arg}) -> {31, [Handle, Func, UniqueId, Arg]};
pack_command(all_yours, {}) -> {24, []};
pack_command(noop, {}) -> {6, []};
pack_command(no_job, {}) -> {10, []};
pack_command(submit_job, {Func, Uniq, Arg}) -> {7, [Func, Uniq, Arg]};
pack_command(submit_job_high, {Func, Uniq, Arg}) -> {21, [Func, Uniq, Arg]};
pack_command(submit_job_high_bg, {Func, Uniq, Arg}) -> {32, [Func, Uniq, Arg]};
pack_command(submit_job_low, {Func, Uniq, Arg}) -> {33, [Func, Uniq, Arg]};
pack_command(submit_job_low_bg, {Func, Uniq, Arg}) -> {34, [Func, Uniq, Arg]};
pack_command(submit_job_bg, {Func, Uniq, Arg}) -> {18, [Func, Uniq, Arg]};
pack_command(job_created, {Handle}) -> {8, [Handle]};
pack_command(get_status, {Handle}) -> {15, [Handle]};
pack_command(option_req, {Option}) -> {26, [Option]};
pack_command(option_res, {Option}) -> {27, [Option]};
pack_command(status_res, {Handle, Known, Running, Numerator, Denominator}) -> {20, [Handle, integer_to_list(Known), integer_to_list(Running), integer_to_list(Numerator), integer_to_list(Denominator)]}.


%% Join a list of lists into a single list separated by Separator
join([], _) -> [];
join([Head|Lists], Separator) ->
     lists:flatten([Head | [[Separator, Next] || Next <- Lists]]).


%% Split a list into multiple lists by Separator
split(List, Separator, Count) ->
    split(List, Separator, [], [], Count).

split([], _, [], [], _) -> [];  %% empty set
split([], _, Lists, Current, _) -> Lists ++ [Current]; %% nothing left to split
split(List, _, Lists, [], 0) -> Lists ++ [List];
split([Separator|Rest], Separator, Lists, Current, Count) when Count > 0 ->
    split(Rest, Separator, Lists ++ [Current], [], Count-1);
split([Other|Rest], Separator, Lists, Current, Count) when Count > 0 ->
    split(Rest, Separator, Lists, Current ++ [Other], Count).

%% handling the logic
process_request(Bin, Msgs) ->
    case protocol:parse_command(Bin) of
        {ok,<<>>, _Kind, {Cmd,Arg1, Arg2}} ->
            lists:reverse([{Cmd,[Arg1, Arg2]} | Msgs]) ;

        {ok,LeftOver, _Kind, {Cmd,Arg1, Arg2}} ->
            process_request(LeftOver, [{Cmd,[Arg1, Arg2]} | Msgs]) ;

        {ok,<<>>, _Kind, {Cmd,Arg1, Arg2, Arg3}} ->
            lists:reverse([{Cmd,[Arg1, Arg2, Arg3]} | Msgs]) ;

        {ok,LeftOver, _Kind, {Cmd,Arg1, Arg2, Arg3}} ->
            process_request(LeftOver, [{Cmd,[Arg1, Arg2, Arg3]} | Msgs]) ;

        {ok,<<>>, _Kind, {Cmd,Args}} ->
            lists:reverse([{Cmd,Args} | Msgs]) ;

        {ok, LeftOver, _Kind, {Cmd,Args}} ->
            process_request(LeftOver, [{Cmd,Args} | Msgs]) ;

        {ok,<<>>, _Kind, Cmd} ->
            lists:reverse([{Cmd, none} | Msgs]) ;

        {ok, LeftOver, _Kind, Cmd} ->
            process_request(LeftOver, [{Cmd, none} | Msgs]) ;

        Error ->
            {error, Error}
    end .


%% tests
process_request_test() ->
    Cmd = <<0,82,69,81,0,0,0,22,0,0,0,0,0,82,69,81,0,0,0,1,0,0,0,19,100,105,
              114,101,99,116,111,114,121,95,112,105,110,103,95,97,108,102,97,
              0,82,69,81,0,0,0,1,0,0,0,19,100,105,114,101,99,116,111,114,121,
              95,101,120,105,116,95,97,108,102,97,0,82,69,81,0,0,0,1,0,0,0,26,
              100,105,114,101,99,116,111,114,121,95,103,101,110,101,114,97,
              116,111,114,95,119,111,114,107,101,114,0,82,69,81,0,0,0,9,0,0,0,
              0>>,
    Result = process_request(Cmd,[]),
    ?assertEqual(5, length(Result)) .
